### ISSUE TITLE
How come C++ cold-starts are so slow? 

### ISSUE THREAD

**Issue by V4G4X** [maxday/lambda-perf#711]
I understand rust and go enjoying fast cold starts like:
<img width="638" alt="image" src="https://github.com/maxday/lambda-perf/assets/34249137/47661771-1d7f-400c-9d5b-9bca51f747a6">

But what could be the reason behind C++ being this slow on the same runtime?
<img width="312" alt="image" src="https://github.com/maxday/lambda-perf/assets/34249137/00d4f27c-6a98-4a10-a771-fcca5ae2c9c2">

Does it have to do with Rust and Go binaries being built with the standard tools? [cargo](https://github.com/maxday/lambda-perf/raw/main/s3-uploader/runtimes/rust_on_provided_al2/Dockerfile) and [go](https://github.com/maxday/lambda-perf/raw/main/s3-uploader/runtimes/go_on_provided_al2/Dockerfile)?  
While the C++ uses an external build tool like [aws-lambda-cpp](https://github.com/awslabs/aws-lambda-cpp)? 

Or is Lambda just less optimized for C++ given it's not used as much?

**Comment by marcomagdy** [comment]
Original author of the Lambda C++ runtime here.

The issue is that @maxday is building the C++ runtime with [a container running Arch Linux](https://github.com/maxday/lambda-perf/blob/01ebbf5921fcea44f7ca1ad2c9ac3758edf0066d/s3-uploader/runtimes/cpp11/Dockerfile#L2), therefore the C++ Lambda runtime adds the entire libc runtime to the final package to ensure it runs correctly on Amazon Linux.

If @maxday builds the C++ runtime using the Amazon Linux container ([just like he is building the Rust runtime](https://github.com/maxday/lambda-perf/blob/01ebbf5921fcea44f7ca1ad2c9ac3758edf0066d/s3-uploader/runtimes/rust_on_provided_al2/Dockerfile#L1)), then the libc runtime does not need to be packaged with his code, and the resulting zip archive will be much smaller. That will reflect on the cold startup time (because Lambda must download the zip archive from S3 on cold start).

In the C++ runtime README, I explained how to avoid packaging the C runtime with your code.

>You can avoid packaging the C runtime in your zip file. This can be done by passing the NO_LIBC flag in CMake as follows:
`aws_lambda_package_target(${PROJECT_NAME} NO_LIBC)
`

Again, this is OK if you build your Lambda function on Amazon Linux (which is the only environment Rust supports btw).

Notably, you will see that C++ is faster than Rust in warm invocations, because there is no big packages to download then.

So, to summarize, if you make the following changes, C++ will be as fast as Rust in cold-starts:
1. Build the container using Amazon Linux container (the same container used today to build the Rust runtime).
2. In your [Lambda function CMake file](https://github.com/maxday/lambda-perf/blob/01ebbf5921fcea44f7ca1ad2c9ac3758edf0066d/s3-uploader/runtimes/cpp11/lambda/CMakeLists.txt#L11) pass an additional argument `NO_LIBC`


@maxday it would be nice to incorporate those change and see the results.

**Comment by maxday** [comment]
Awesome! thanks you so much @marcomagdy and @V4G4X!
I'll definitely update this over the week-end! I'll keep you posted!
